apisix/plugins/grpc-web.lua

```diff
@@ -147,6 +151,49 @@ function _M.body_filter(conf, ctx)
         chunk = encode_base64(chunk)
         ngx_arg[1] = chunk
     end
+
+    --[[
+    upstream_trailer_* available since NGINX version 1.13.10 :
+    https://nginx.org/en/docs/http/ngx_http_upstream_module.html#var_upstream_trailer_
+
+    grpc-web trailer format reference:
+    envoyproxy/envoy/source/extensions/filters/http/grpc_web/grpc_web_filter.cc
+
+    Format for grpc-web trailer
+        1 byte: 0x80
+        4 bytes: length of the trailer
+        n bytes: trailer
+
+    --]]
+    local status = ctx.var.upstream_trailer_grpc_status
+    if status ~= "" and status ~= nil then
+        local status_str = "grpc-status:" .. tonumber(status)
+        local status_msg = "grpc-message:" .. ctx.var.upstream_trailer_grpc_message
+        local grpc_web_trailer = status_str .. "\r\n" .. status_msg .. "\r\n"
+        local len = #grpc_web_trailer
+
+        -- 1 byte: 0x80
+        local trailer_buf = string.char(0x80)
+        -- 4 bytes: length of the trailer
+        trailer_buf = trailer_buf .. string.char(
+            bit.band(bit.rshift(len, 24), 0xff),
+            bit.band(bit.rshift(len, 16), 0xff),
+            bit.band(bit.rshift(len, 8), 0xff),
+            bit.band(len, 0xff)
+        )
+        -- n bytes: trailer
+        trailer_buf = trailer_buf .. grpc_web_trailer
+
+        if ctx.grpc_web_encoding == CONTENT_ENCODING_BINARY then
```

Comment:
checking the encoding twice, can this be removed now? https://github.com/apache/apisix/blob/196461e2ec728a763ca13cbef337d9ce0fb2abc1/apisix/plugins/grpc-web.lua#L148C1-L153C8

Reply:
This encoding check is for trailer only, the previous check is to base64encode the body (needed only if encoding is base64). This check is required as trailers is needed for both the cases and requires different handling strategy.

This executes only if the status is not pty or nil. The previous check executes every time for the body content.